rust/src/extensions.rs

```diff
@@ -0,0 +1,191 @@
+//! Core logic for extensions.yaml file.
+
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ */
+
+use anyhow::{bail, Context, Result};
+use openat_ext::OpenatDirExt;
+use serde_derive::Deserialize;
+use std::collections::HashMap;
+use std::io::Read;
+
+use crate::ffi::StringMapping;
+use crate::utils;
+
+const RPMOSTREE_EXTENSIONS_STATE_FILE: &str = ".rpm-ostree-state-chksum";
+
+#[derive(Deserialize, Debug)]
+pub struct Extensions {
+    extensions: HashMap<String, Extension>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Extension {
+    packages: Vec<String>,
+    architectures: Option<Vec<String>>,
+    #[serde(rename = "match-base-evr")]
+    match_base_evr: Option<String>,
+}
+
+fn extensions_load_stream(
+    stream: &mut impl std::io::Read,
+    basearch: &str,
+    base_pkgs: &Vec<StringMapping>,
+) -> Result<Box<Extensions>> {
+    let mut parsed: Extensions = serde_yaml::from_reader(stream)?;
+
+    parsed.extensions.retain(|_, ext| {
+        ext.architectures
+            .as_ref()
+            .map(|v| v.iter().any(|a| a == basearch))
+            .unwrap_or(true)
+    });
+
+    let base_pkgs: HashMap<String, String> =
```

Comment:
I think this could be a `HashMap<&str, &str>` at least and point into the `base_pkgs` vec to avoid string duplication.